plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java

```diff
@@ -443,6 +464,58 @@ private static <K, V> Map<K, V> getAll(Cache<K, AtomicReference<V>> cache, Itera
         return result.buildOrThrow();
     }
 
+    private static <K, V> Map<K, V> getAll(
+            Cache<K, AtomicReference<V>> cache,
+            Iterable<K> keys,
+            Function<Collection<K>, Map<K, V>> bulkLoader,
+            Predicate<V> isSufficient,
+            BiFunction<V, V, V> merger)
+    {
+        ImmutableMap.Builder<K, V> result = ImmutableMap.builder();
+        Map<K, AtomicReference<V>> toLoad = new HashMap<>();
+
+        keys.forEach(key -> {
+            // make sure the value holder is retrieved before the new values are loaded
+            // so that in case of invalidation we will not set the stale value
+            AtomicReference<V> currentValueHolder = uncheckedCacheGet(cache, key, AtomicReference::new);
+            V currentValue = currentValueHolder.get();
+            if (currentValue != null && isSufficient.test(currentValue)) {
+                result.put(key, currentValue);
+            }
+            else {
+                toLoad.put(key, currentValueHolder);
+            }
+        });
+
+        if (toLoad.isEmpty()) {
+            return result.buildOrThrow();
+        }
+
+        Map<K, V> newEntries = bulkLoader.apply(toLoad.keySet());
+        toLoad.forEach((key, valueHolder) -> {
+            V newValue = newEntries.get(key);
+            if (newValue == null) {
+                throw new InvalidCacheLoadException("Failed to return a value for " + key);
+            }
+            result.put(key, merge(merger, valueHolder, valueHolder.get(), newValue));
```

Comment:
i realized we don't need to return the merged value, since the newValue is fully loaded.

so we can go without merge
```
toLoad.forEach((key, valueHolder) -> {
    V newValue = newEntries.get(key);
    verifyNotNull(newValue, "no value for %s", key);
    valueHolder.accumulateAndGet(newValue, merger);
    result.put(key, newValue); // newValue is fully loaded, not incremental
});
```

Reply:
I think there is a race here with 2 concurrent loads so merge is necessary

Reply:
never mind, you used merger in your code snippet